### PR TITLE
Add support for context-receivers

### LINKED ISSUE #1232: Add support for context receivers
Kotlin 1.6.20 has added [context receivers](https://github.com/Kotlin/KEEP/blob/master/proposals/context-receivers.md) however there is currently no way to generate them using kotlin poet.

These would be needed for both lambdas and functions

### PR DESCRIPTION
**Pull request by DRSchlaubi**
This PR aims at implementing support for context receivers (#1232)

It currently only tries to implement a possible design for builders and specs and has no code generation logic

```kotlin
@OptIn(ContextReceivers::class)
public fun main() {
  val contextFunction = FunSpec.Builder("withContext")
    .contextReceiver(InputStream::class)
    .build()

  val contextLambda = LambdaTypeName.get(
    receiver = STRING,
    contextReceivers = listOf(InputStream::class.asTypeName()),
    returnType = STRING
  )
}
```

Fixes #1232

### CODE REVIEW

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt [hunk 1]
```diff
@@ -13,13 +13,10 @@
...
  */
+@file:OptIn(ContextReceivers::class)
 package com.squareup.kotlinpoet
 
-import com.squareup.kotlinpoet.KModifier.ABSTRACT
-import com.squareup.kotlinpoet.KModifier.EXPECT
-import com.squareup.kotlinpoet.KModifier.EXTERNAL
-import com.squareup.kotlinpoet.KModifier.INLINE
-import com.squareup.kotlinpoet.KModifier.VARARG
+import com.squareup.kotlinpoet.KModifier.*
```
Comment (ZacSweers) [inline kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt:19]: Please abide by our code styles

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/ContextReceivers.kt [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package com.squareup.kotlinpoet
+
+/**
+ * Annotation marking APIs for the [KEEP-259](https://github.com/Kotlin/KEEP/blob/master/proposals/context-receivers.md)
+ * prototype in Kotlin 1.6.20
+ */
+@RequiresOptIn
+@MustBeDocumented
+@Target(AnnotationTarget.CLASS, AnnotationTarget.FUNCTION, AnnotationTarget.VALUE_PARAMETER, AnnotationTarget.PROPERTY)
+public annotation class ContextReceivers
```
Comment (ZacSweers) [inline kotlinpoet/src/main/java/com/squareup/kotlinpoet/ContextReceivers.kt:10]: We have an experimental API annotation already, please reuse that

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt [hunk 3]
```diff
@@ -368,7 +391,8 @@ public class FunSpec private constructor(
       this.receiverKdoc = kdoc
     }
 
-    @JvmOverloads public fun receiver(
+    @JvmOverloads
```
Comment (ZacSweers) [inline kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt:394]: Please clean up these unrelated codestyle changes

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/LambdaTypeName.kt [hunk 4]
```diff
@@ -17,8 +17,11 @@ package com.squareup.kotlinpoet
 
 import kotlin.reflect.KClass
 
+@OptIn(ContextReceivers::class)
 public class LambdaTypeName private constructor(
   public val receiver: TypeName? = null,
+  @property:ContextReceivers
```
Comment (ZacSweers) [inline kotlinpoet/src/main/java/com/squareup/kotlinpoet/LambdaTypeName.kt:23]: Why the site target?

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/LambdaTypeName.kt [hunk 5]
```diff
@@ -17,8 +17,11 @@ package com.squareup.kotlinpoet
...
+  public val contextReceivers: List<TypeName>? = emptyList(),
```
Comment (ZacSweers) [inline kotlinpoet/src/main/java/com/squareup/kotlinpoet/LambdaTypeName.kt:24]: We should pick null or empty but not both

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt [hunk 6]
```diff
@@ -359,7 +359,30 @@ public class FunSpec private constructor(
       typeVariables += typeVariable
     }
 
-    @JvmOverloads public fun receiver(
+    public fun contextReceiver(receiverTypes: Collection<TypeName>): Builder = apply {
```
Comment (ZacSweers) [inline kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt:362]: This is too many overloads IMO. I think we should stick with just two

```
addConextReceiver(TypeName)
addConextReceivers(Iterable<TypeName>)
```

**Review by ZacSweers — COMMENTED** [review]
Are you open to implementing the code gen part of it? I don't think it would be too difficult.

General design looks ok, left comments where there need to be changes.

In the future, please don't PR unrelated formatting changes that deviate from the code style we use, it makes the PR quite noisy.

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/LambdaTypeName.kt (hunk 5 above)
Comment (DRSchlaubi) [reply kotlinpoet/src/main/java/com/squareup/kotlinpoet/LambdaTypeName.kt:24]: Looks like an oversight on my part, it was supposed to be not nullable and default to empty

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/LambdaTypeName.kt (hunk 4 above)
Comment (DRSchlaubi) [reply kotlinpoet/src/main/java/com/squareup/kotlinpoet/LambdaTypeName.kt:23]: IntelliJ gave me an error that the annotation wouldn't work on parameters, despite it also having the `VALUE_PARAMETER` target

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt (hunk 6 above)
Comment (DRSchlaubi) [reply kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt:362]: the receiver function has 3 overloads as well, I would prefer to have at least the vararg overloads

Comment (ZacSweers) [reply kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt:362]: I'm ok with vararg 👍

**Comment by DRSchlaubi** [comment]
> Are you open to implementing the code gen part of it? I don't think it would be too difficult.

I was going to propose an implementation with this design as well; however, I didn't I ran into an issue, which is why I just proposed a design, once that is finished I would be open to implement the code gen

**Comment by DRSchlaubi** [comment]
I hope I caught all the code style violations?

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt [hunk 7]
```diff
@@ -45,6 +46,9 @@ public class FunSpec private constructor(
   public val modifiers: Set<KModifier> = builder.modifiers.toImmutableSet()
   public val typeVariables: List<TypeVariableName> = builder.typeVariables.toImmutableList()
   public val receiverType: TypeName? = builder.receiverType
+
+  @DelicateKotlinPoetApi("Context receivers are currently a preview feature")
```
Comment (ZacSweers) [inline kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt:50]: Hmmm, this isn't the right one either. I thought we had an existing experimental annotation API (`@ExperimentalKotlinPoetApi` or similar). If there isn't, please make one and use that

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt [hunk 8]
```diff
@@ -359,6 +368,23 @@ public class FunSpec private constructor(
       typeVariables += typeVariable
     }
 
+    @DelicateKotlinPoetApi("Context receivers are currently a preview feature")
+    public fun contextReceiver(receiverTypes: Collection<TypeName>): Builder = apply {
+      check(!name.isConstructor) { "$name cannot have receiver type" }
```
Comment (ZacSweers) [inline kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt:373]:
```suggestion
      check(!name.isConstructor) { "$name: constructors cannot have context receivers" }
```

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt [hunk 9]
```diff
@@ -359,6 +368,23 @@ public class FunSpec private constructor(
...
+      contextReceiverTypes += receiverTypes
+    }
+
+    @DelicateKotlinPoetApi("Context receivers are currently a preview feature")
+    public fun contextReceiver(vararg receiverType: TypeName): Builder = contextReceiver(receiverType.toList())
+
+    @DelicateKotlinPoetApi("Context receivers are currently a preview feature")
+    public fun contextReceiver(vararg receiverType: Type): Builder =
```
Comment (ZacSweers) [inline kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt:381]: Remove this and the KClass overload below this and _only_ expose TypeName overloads. I think you misunderstood my previous response. We should have two - a vararg param and an Iterable plural overload

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/LambdaTypeName.kt [hunk 10]
```diff
@@ -80,12 +87,20 @@ public class LambdaTypeName private constructor(
   }
 
   public companion object {
+    /** Returns a lambda type with `returnType` and parameters listed in `parameters`. */
+    @DelicateKotlinPoetApi("Context receivers are currently a preview feature") @JvmStatic public fun get(
+      receiver: TypeName? = null,
+      contextReceivers: List<TypeName> = emptyList(),
```
Comment (ZacSweers) [inline kotlinpoet/src/main/java/com/squareup/kotlinpoet/LambdaTypeName.kt:93]: While I get why we can't just add this to the below one, let's at least add this argument to the end of the parameters list

File: kotlinpoet/src/test/java/com/squareup/kotlinpoet/FunSpecTest.kt [hunk 11]
```diff
@@ -429,6 +429,20 @@ class FunSpecTest {
...
 
+  @Test fun functionWithContextReceiver() {
+    val stringType = STRING
+    val funSpec = FunSpec.builder("foo")
+      .contextReceiver(stringType)
+      .build()
+
+    assertThat(funSpec.toString()).isEqualTo(
+      """
+      |context(kotlin.String) public fun foo(): kotlin.Unit {
```
Comment (ZacSweers) [inline kotlinpoet/src/test/java/com/squareup/kotlinpoet/FunSpecTest.kt:440]: The receiver should be the line above the function right?

Also please add more thorough tests for this:
- check that it fails on constructors
- add multiple types
- add generic types (can these be generic and reference generic types in the function descriptor?)
- add a typevariablename (or if that's not legal, add a check + test)
- typenames with annotations (or a check + test if that's not legal)

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/LambdaTypeName.kt [hunk 12]
```diff
@@ -17,8 +17,11 @@ package com.squareup.kotlinpoet
 
 import kotlin.reflect.KClass
 
+@OptIn(DelicateKotlinPoetApi::class)
 public class LambdaTypeName private constructor(
   public val receiver: TypeName? = null,
+  @property:DelicateKotlinPoetApi("Context receivers are currently a preview feature")
+  public val contextReceivers: List<TypeName> = emptyList(),
```
Comment (ZacSweers) [inline kotlinpoet/src/main/java/com/squareup/kotlinpoet/LambdaTypeName.kt:24]: Is this the right place for this, vs a TypeName property like annotations are?

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt (hunk 7 above)
Comment (DRSchlaubi) [reply kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt:50]: I thought it wasn't the right one, but even though there are Experimental/Preview annotations these are for other modules like the ksp interop

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt (hunk 8 above)
Comment (DRSchlaubi) [reply kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt:373]: This was copied from the receiver function, so then we would have two error messages for the same thing?

Comment (DRSchlaubi) [reply kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt:373]: Ok fine then I'll change it

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/LambdaTypeName.kt (hunk 12 above)
Comment (ZacSweers) [reply kotlinpoet/src/main/java/com/squareup/kotlinpoet/LambdaTypeName.kt:24]: Hmm, I suppose so. I guess we can start with this and can always change it later if need be while it's still experimental

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt [hunk 13]
```diff
@@ -84,6 +88,10 @@ public class FunSpec private constructor(
       codeWriter.emitKdoc(kdoc.ensureEndsWithNewLine())
     }
     codeWriter.emitAnnotations(annotations, false)
+    if (contextReceiverTypes.isNotEmpty()) {
+      val receivers = contextReceiverTypes.joinToString(", ") { "%T" }
+      codeWriter.emitCode("context($receivers)\n", *contextReceiverTypes.toTypedArray())
```
Comment (Egorand) [inline kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt:93]: nit: alternative version
```suggestion
      val receivers = contextReceiverTypes
        .map { CodeBlock.of("%T", it) }
        .joinToCode(prefix = "context(", suffix = ")\n"))
      codeWriter.emitCode(receivers)
```

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt [hunk 14]
```diff
@@ -359,6 +368,15 @@ public class FunSpec private constructor(
       typeVariables += typeVariable
     }
 
+    @ExperimentalKotlinPoetApi
+    public fun contextReceiver(receiverTypes: Collection<TypeName>): Builder = apply {
```
Comment (Egorand) [inline kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt:372]: Should this be `contextReceivers` (multiple)?

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt [hunk 15]
```diff
@@ -359,6 +368,15 @@ public class FunSpec private constructor(
...
+      check(!name.isConstructor) { "$name: constructors cannot have context receivers" }
+      contextReceiverTypes += receiverTypes
+    }
+
+    @ExperimentalKotlinPoetApi
+    public fun contextReceiver(vararg receiverType: TypeName): Builder = contextReceiver(receiverType.toList())
```
Comment (Egorand) [inline kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt:378]: Same here

File: kotlinpoet/src/test/java/com/squareup/kotlinpoet/FunSpecTest.kt [hunk 16]
```diff
@@ -429,6 +432,76 @@ class FunSpecTest {
...
+    assertThat(funSpec.toString()).isEqualTo(
+      """
+      |context(kotlin.String)
+      |public fun foo(): kotlin.Unit {
+      |}
+      |""".trimMargin()
+    )
+  }
+
+  @Test fun functionWithMultipleContextReceiver() {
```
Comment (Egorand) [inline kotlinpoet/src/test/java/com/squareup/kotlinpoet/FunSpecTest.kt:450]:
```suggestion
  @Test fun functionWithMultipleContextReceivers() {
```

File: kotlinpoet/src/test/java/com/squareup/kotlinpoet/FunSpecTest.kt [hunk 17]
```diff
@@ -429,6 +432,76 @@ class FunSpecTest {
...
+    val stringType = STRING
```
Comment (Egorand) [inline kotlinpoet/src/test/java/com/squareup/kotlinpoet/FunSpecTest.kt:451]: nit: are these variables necessary? Feels like they're easy to inline.

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt [hunk 18]
```diff
@@ -359,6 +368,15 @@ public class FunSpec private constructor(
       typeVariables += typeVariable
     }
 
+    @ExperimentalKotlinPoetApi
+    public fun contextReceiver(receiverTypes: Collection<TypeName>): Builder = apply {
+      check(!name.isConstructor) { "$name: constructors cannot have context receivers" }
```
Comment (Egorand) [inline kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt:373]: Will name here ever be different from `constructor`? If no - should we omit it from the error message?

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/LambdaTypeName.kt [hunk 19]
```diff
@@ -50,10 +53,14 @@ public class LambdaTypeName private constructor(
     suspending: Boolean = this.isSuspending,
     tags: Map<KClass<*>, Any> = this.tags.toMap()
   ): LambdaTypeName {
-    return LambdaTypeName(receiver, parameters, returnType, nullable, suspending, annotations, tags)
+    return LambdaTypeName(receiver, contextReceivers, parameters, returnType, nullable, suspending, annotations, tags)
   }
 
   override fun emit(out: CodeWriter): CodeWriter {
+    if (contextReceivers.isNotEmpty()) {
```
Comment (Egorand) [inline kotlinpoet/src/main/java/com/squareup/kotlinpoet/LambdaTypeName.kt:60]: Can maybe introduce a `CodeWriter.emitContextReceivers` method to avoid duplication.

File: kotlinpoet/src/test/java/com/squareup/kotlinpoet/LambdaTypeNameTest.kt [hunk 20]
```diff
@@ -51,6 +52,58 @@ class LambdaTypeNameTest {
     )
   }
 
+  @Test fun contextReceiver() {
+    val typeName = LambdaTypeName.get(
+      Int::class.asTypeName(),
```
Comment (Egorand) [inline kotlinpoet/src/test/java/com/squareup/kotlinpoet/LambdaTypeNameTest.kt:57]: Argument names would be nice to have here for better readability.

File: kotlinpoet/src/test/java/com/squareup/kotlinpoet/LambdaTypeNameTest.kt [hunk 21]
```diff
@@ -51,6 +52,58 @@ class LambdaTypeNameTest {
...
+      listOf(),
+      Unit::class.asTypeName(),
+      listOf(STRING)
+    )
+    assertThat(typeName.toString()).isEqualTo(
+      "context(kotlin.String) kotlin.Int.() -> kotlin.Unit"
+    )
+  }
+
+  @Test fun functionWithMultipleContextReceiver() {
```
Comment (Egorand) [inline kotlinpoet/src/test/java/com/squareup/kotlinpoet/LambdaTypeNameTest.kt:67]:
```suggestion
  @Test fun functionWithMultipleContextReceivers() {
```

**Review by Egorand — APPROVED** [review]
Sweet! Added a few comments, some of them (non-nit ones) would be nice to address before merging.

**Comment by Egorand** [comment]
@DRSchlaubi also please sign our [CLA](https://spreadsheets.google.com/spreadsheet/viewform?formkey=dDViT2xzUHAwRkI3X3k5Z0lQM091OGc6MQ&ndplr=1) before we can merge you PR.

**Comment by DRSchlaubi** [comment]
Done

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt (hunk 18 above)
Comment (DRSchlaubi) [reply kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt:373]: I was asking myself the same question, but this is consistent with receiver and @ZacSweers suggested it

File: kotlinpoet/src/test/java/com/squareup/kotlinpoet/FunSpecTest.kt (hunk 17 above)
Comment (Egorand) [reply kotlinpoet/src/test/java/com/squareup/kotlinpoet/FunSpecTest.kt:451]: Makes sense!

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt (hunk 18 above)
Comment (Egorand) [reply kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt:371]: Got it, sounds good!

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt [hunk 22]
```diff
@@ -359,6 +366,15 @@ public class FunSpec private constructor(
       typeVariables += typeVariable
     }
 
+    @ExperimentalKotlinPoetApi
+    public fun contextReceivers(receiverTypes: Collection<TypeName>): Builder = apply {
```
Comment (Egorand) [inline kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt:370]: Should we use `Iterable<TypeName>` for consistency with other methods? It's also a less-specific type.
```suggestion
    public fun contextReceivers(receiverTypes: Iterable<TypeName>): Builder = apply {
```

**Review by Egorand — APPROVED** [review]

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt (hunk 22 above)
Comment (DRSchlaubi) [reply kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt:370]: Doesn't addAll take a collection?

Comment (Egorand) [reply kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt:370]: There's an override that takes an `Iterable`, most methods in `FunSpec.Builder` use it.

**Comment by Egorand** [comment]
Think it just needs a `./gradlew :kotlinpoet:spotlessApply`.

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/LambdaTypeName.kt [hunk 23]
```diff
@@ -50,10 +53,12 @@ public class LambdaTypeName private constructor(
     suspending: Boolean = this.isSuspending,
     tags: Map<KClass<*>, Any> = this.tags.toMap()
   ): LambdaTypeName {
-    return LambdaTypeName(receiver, parameters, returnType, nullable, suspending, annotations, tags)
+    return LambdaTypeName(receiver, contextReceivers, parameters, returnType, nullable, suspending, annotations, tags)
   }
 
   override fun emit(out: CodeWriter): CodeWriter {
+    out.emitContextReceivers(contextReceivers)
+    out.emit(" ")
```
Comment (ZacSweers) [inline kotlinpoet/src/main/java/com/squareup/kotlinpoet/LambdaTypeName.kt:61]: should this be with a space-preserving dot?

```suggestion
    out.emit("·")
```

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt (hunk 14 above)
[the remark this answers is not in the document]
Comment (Egorand) [reply kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt]: Already fixed.

File: kotlinpoet/src/test/java/com/squareup/kotlinpoet/FunSpecTest.kt [hunk 24]
```diff
@@ -429,6 +432,76 @@ class FunSpecTest {
...
+  @Test fun functionWithGenericContextReceiver() {
+    val genericType = TypeVariableName("T")
+    val funSpec = FunSpec.builder("foo")
+      .addTypeVariable(genericType)
+      .contextReceiver(genericType)
+      .build()
+
+    assertThat(funSpec.toString()).isEqualTo(
+      """
+      |context(T)
```
[the remark this answers is not in the document]
Comment (DRSchlaubi) [reply kotlinpoet/src/test/java/com/squareup/kotlinpoet/FunSpecTest.kt:476]: It is yes

File: kotlinpoet/src/test/java/com/squareup/kotlinpoet/FunSpecTest.kt [hunk 25]
```diff
@@ -429,6 +432,76 @@ class FunSpecTest {
...
+
+  @Test fun functionWithAnnotatedContextReceiver() {
+    val genericType = STRING.copy(annotations = listOf(AnnotationSpec.get(TestAnnotation())))
+    val funSpec = FunSpec.builder("foo")
+      .contextReceiver(genericType)
+      .build()
+
+    assertThat(funSpec.toString()).isEqualTo(
+      """
+      |context(@com.squareup.kotlinpoet.FunSpecTest.TestAnnotation kotlin.String)
```
[the remark this answers is not in the document]
Comment (DRSchlaubi) [reply kotlinpoet/src/test/java/com/squareup/kotlinpoet/FunSpecTest.kt:491]: Using it in IntelliJ crashes intellij, however there is no real rules in the KEEP for how to use annotations and the compiler accepts it

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/LambdaTypeName.kt (hunk 23 above)
Comment (DRSchlaubi) [reply kotlinpoet/src/main/java/com/squareup/kotlinpoet/LambdaTypeName.kt:61]: yes I missed that

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt [hunk 26]
```diff
@@ -84,6 +88,7 @@ public class FunSpec private constructor(
       codeWriter.emitKdoc(kdoc.ensureEndsWithNewLine())
     }
     codeWriter.emitAnnotations(annotations, false)
+    codeWriter.emitContextReceivers(contextReceiverTypes, "\n")
```
Comment (Egorand) [inline kotlinpoet/src/main/java/com/squareup/kotlinpoet/FunSpec.kt:91]: nit:
```suggestion
    codeWriter.emitContextReceivers(contextReceiverTypes, suffix = "\n")
```

File: kotlinpoet/src/main/java/com/squareup/kotlinpoet/LambdaTypeName.kt [hunk 27]
```diff
@@ -50,10 +53,11 @@ public class LambdaTypeName private constructor(
     suspending: Boolean = this.isSuspending,
     tags: Map<KClass<*>, Any> = this.tags.toMap()
   ): LambdaTypeName {
-    return LambdaTypeName(receiver, parameters, returnType, nullable, suspending, annotations, tags)
+    return LambdaTypeName(receiver, contextReceivers, parameters, returnType, nullable, suspending, annotations, tags)
   }
 
   override fun emit(out: CodeWriter): CodeWriter {
+    out.emitContextReceivers(contextReceivers, "·")
```
Comment (Egorand) [inline kotlinpoet/src/main/java/com/squareup/kotlinpoet/LambdaTypeName.kt:60]: nit:
```suggestion
    out.emitContextReceivers(contextReceivers, suffix = "·")
```

**Review by Egorand — APPROVED** [review]

**Comment by Egorand** [comment]
Merged, thank you!
